osmomath/decimal.go

```diff
@@ -552,23 +564,46 @@ func (d BigDec) MustFloat64() float64 {
 // Dec returns the osmomath.Dec representation of a BigDec.
 // Values in any additional decimal places are truncated.
 func (d BigDec) Dec() Dec {
-	precisionDiff := PrecisionBigDec - PrecisionDec
-	precisionFactor := new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(precisionDiff)), nil)
+	return d.DecWithPrecision(sdk.Precision)
+}
 
-	if precisionDiff < 0 {
-		panic("invalid decimal precision")
+func (d BigDec) DecWithPrecision(precision int64) Dec {
+	var precisionFactor *big.Int
+	if precision >= sdk.Precision {
```

Comment:
What if `precision` is 19? Can we account for that?

Also, would be good to have a check making sure that we never exceed `BigDec` precision of 36. If that happens, panic

Reply:
> What if precision is 19? Can we account for that?

this is a matter of desired design. here, if a precision is 19, it will proceed as if the precision is as big as `Dec` allows it to be (18)

> Also, would be good to have a check making sure that we never exceed BigDec precision of 36.

good point, will add

Reply:
> this is a matter of desired design. here, if a precision is 19, it will proceed as if the precision is as big as Dec allows it to be (18)

Ah right! I confused this method with ChopPrecision for `BigDec`

Reply:
hey @p0mvn! On the other thought, many functions in `decimal.go` assume that precision provided is less than `PrecisionBigDec` (keyword for search: "CONTRACT: prec <="). I suppose it is for performance purposes to avoid checking provided precision every time operation is performed. I think it makes sense to follow the same pattern with `ChopPrecision` functions, wdyt?

Reply:
also:

> What if precision is 19? Can we account for that?

> this is a matter of desired design. here, if a precision is 19, it will proceed as if the precision is as big as Dec allows it to be (18)

every call to `DecWithPrecision` function would make this check whether a value is not bigger than `PrecisionDec`. Maybe we also keep it a function user's task to make sure provided precision does not exceed `PrecisionDec`? for performance purposes

Reply:
>I suppose it is for performance purposes to avoid checking provided precision every time operation is performed. 

I think that the added safety benefits outweigh any performance gains from this. Modern compilers should be able to optimize such checks pretty well so I think it makes sense to the check